### PR TITLE
ggml : multi-thread ggml_rope() (~3-4 times faster on M1)

### PR DESCRIPTION
**Pull request by ggerganov**
### Summary
🧶🚀🧑‍💻

This pull request improves the performance of the `ggml` library by parallelizing the rope operation on tensors, and modifying the graph executor to handle parallel tasks. It affects the file `ggml.c`.

> _We'll hoist the sail with a rope operation_
> _`ggml_compute_forward_rope_f32` and `f16`_
> _We'll work in parallel, no more hesitation_
> _On the count of three, pull hard, me friends_

### Walkthrough
*  Parallelize the rope operation for both `f32` and `f16` data types by dividing the input rows among the available threads ([link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7241), [link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7268-R7288), [link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7298), [link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7325-R7361))
*  Remove the invalid assertions that the thread index is zero from the `ggml_compute_forward_rope_f32` and `ggml_compute_forward_rope_f16` functions ([link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7241), [link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL7298))
*  Set the number of tasks for the rope operation node to the number of threads in the `ggml.c` file ([link](https://github.com/ggerganov/llama.cpp/pull/781/files?diff=unified&w=0#diff-6d9ce99fcb6f51ff76f59e479f6e6fc0bb62edef7442805d7a5bb15b23996b5dL9427-R9460))

### CODE REVIEW

**Review by prusnak — APPROVED** [review]
Tested on M1.

Before the change (master 5a8c4f624077373a198cd562146ffa67b02ebc75) - second run:
```
llama_print_timings:        load time =  2294.47 ms
llama_print_timings:      sample time =   116.82 ms /   128 runs   (    0.91 ms per run)
llama_print_timings: prompt eval time =  2090.16 ms /     8 tokens (  261.27 ms per token)
llama_print_timings:        eval time = 23215.24 ms /   127 runs   (  182.80 ms per run)
llama_print_timings:       total time = 25628.04 ms
```

After the change (commit 625f212e38b410832c3af26ba03a3b0013866d9a) - second run:

```
llama_print_timings:        load time =  1460.12 ms
llama_print_timings:      sample time =    94.96 ms /   128 runs   (    0.74 ms per run)
llama_print_timings: prompt eval time =  1272.30 ms /     8 tokens (  159.04 ms per token)
llama_print_timings:        eval time = 21661.06 ms /   127 runs   (  170.56 ms per run)
llama_print_timings:       total time = 23217.51 ms
```

File: ggml.c [hunk 1]
```diff
@@ -7265,11 +7264,28 @@ static void ggml_compute_forward_rope_f32(
...
+
     for (int64_t i3 = 0; i3 < ne3; i3++) {
         for (int64_t i2 = (mode == 0 ? 0 : n_past); i2 < ne2; i2++) {
             const int p = (mode == 0 ? n_past + i2 : i2);
             for (int64_t i1 = 0; i1 < ne1; i1++) {
+                if (ir++ < ir0) continue;
+                if (ir   > ir1) break;
+
                 for (int i0 = 0; i0 < n_dims; i0 += 2) {
                     const float theta = powf(10000.0, ((float)-i0)/n_dims);
```
Comment (howard0su) [inline ggml.c:7290]: theta can be calculated as theta *= factor in each loop. factor can be calculated out of the outer loop factor = powf(10000.0, ((float)-2)/n_dims); the initial theta = p.

Comment (ggerganov) [reply ggml.c:7290]: Open a PR if you observe performance gain improvement

**Comment by fatemehkhoramdel** [comment]
How can i use Multi thread llama on cpu?

**Pull request merged**